subsys/bluetooth/services/ots/ots_client.c

```diff
@@ -169,15 +129,89 @@ static int read_attr(struct bt_conn *conn,
 		    struct bt_otc_internal_instance_t *inst,
 		    uint16_t handle, bt_gatt_read_func_t cb);
 
-static void print_oacp_response(enum bt_ots_oacp_proc_type req_opcode,
-				enum bt_ots_oacp_res_code result_code)
+/* L2CAP callbacks */
+static void tx_done(struct bt_gatt_ots_l2cap *l2cap_ctx,
+		    struct bt_conn *conn)
+{
+	/* Not doing any writes yet */
+	BT_ERR("Unexpected call, context: %p, conn: %p", l2cap_ctx, conn);
+}
+
+static ssize_t rx_done(struct bt_gatt_ots_l2cap *l2cap_ctx,
+		       struct bt_conn *conn, struct net_buf *buf)
+{
+	const uint32_t offset = cur_inst->rcvd_size;
+	bool is_complete = false;
+	const struct bt_otc_obj_metadata *cur_object =
+		&cur_inst->otc_inst->cur_object;
+	int cb_ret;
+
+	BT_DBG("Incoming L2CAP data, context: %p, conn: %p, len: %u, offset: %u",
+	       l2cap_ctx, conn, buf->len, offset);
+
+	cur_inst->rcvd_size += buf->len;
+
+	if (cur_inst->rcvd_size >= cur_object->current_size) {
+		is_complete = true;
+	}
+
+	if (cur_inst->rcvd_size > cur_object->current_size) {
+		BT_WARN("Received %u but expected maximum %u",
+			cur_inst->rcvd_size, cur_object->current_size);
+	}
+
+
+	cb_ret = cur_inst->otc_inst->cb->content_cb(conn, offset,
+						    buf->len, buf->data,
+						    is_complete, 0);
+
+	if (is_complete) {
+		const uint32_t rcv_size = cur_object->current_size;
+		int err;
+
+		BT_DBG("Received the whole object (%u bytes). "
+		       "Disconnecting L2CAP CoC", rcv_size);
+		err = bt_gatt_ots_l2cap_disconnect(l2cap_ctx);
```

Comment:
Is it strictly necessary to closet the L2CAP connection? It would be feasible to use the same connection to read more objects.

Reply:
This is "inherited" code, so I am frankly not sure.
I guess it would be possible to keep it open.  But we do not know neither whether or when it will be needed again, so just as neat to close it.
Besides, the server side closes the other end of the connection anyway.

Reply:
I think the server only closes the connection if there is an error, otherwise it leaves it up to the client to manage the connection. But good point about not knowing when to need it again.

Reply:
We could close it using a timer, but that's outside the scope of this PR. Consider creating an issue for it.